server/src/main/java/org/elasticsearch/transport/InboundDecoder.java

```diff
@@ -203,23 +205,23 @@ private void ensureOpen() {
         }
     }
 
-    static IllegalStateException ensureVersionCompatibility(
-        TransportVersion remoteVersion,
-        TransportVersion currentVersion,
-        boolean isHandshake
-    ) {
-        // for handshakes we are compatible with N-2 since otherwise we can't figure out our initial version
-        // since we are compatible with N-1 and N+1 so we always send our minCompatVersion as the initial version in the
-        // handshake. This looks odd but it's required to establish the connection correctly we check for real compatibility
-        // once the connection is established
-        final TransportVersion compatibilityVersion = isHandshake ? currentVersion.calculateMinimumCompatVersion() : currentVersion;
-        if (remoteVersion.isCompatible(compatibilityVersion) == false) {
-            final TransportVersion minCompatibilityVersion = isHandshake
-                ? compatibilityVersion
-                : compatibilityVersion.calculateMinimumCompatVersion();
-            String msg = "Received " + (isHandshake ? "handshake " : "") + "message from unsupported version: [";
-            return new IllegalStateException(msg + remoteVersion + "] minimal compatible version is: [" + minCompatibilityVersion + "]");
+    static void checkHandshakeVersionCompatibility(TransportVersion handshakeVersion) {
+        if (handshakeVersion.before(TransportHandshaker.EARLIEST_HANDSHAKE_VERSION)) {
+            throw new IllegalStateException(
+                "Received message from unsupported version: ["
+                    + handshakeVersion
+                    + "] minimal compatible version is: ["
+                    + TransportHandshaker.EARLIEST_HANDSHAKE_VERSION
+                    + "]"
+            );
+        }
+    }
+
+    static void checkVersionCompatibility(TransportVersion remoteVersion, TransportVersion currentVersion) {
```

Comment:
Requiring `currentVersion` here is a little weird IMO. I see why it's here: we lie about our current version in a few tests, but I think these are not tests where this check matters; moreover if we do lie about our current version then we may end up claiming to support a transport version which makes no sense to us. Not a topic for this PR I think, but something worth cleaning up independently.

Reply:
I've taken this out as part of https://github.com/elastic/elasticsearch/pull/93453